File: products/workflows/backend/services/wait_clock_conditions.py
"""
Detects `wait_until_condition` conditions that only the polling re-check could ever satisfy:
a comparison against the clock, and a read of a group property.

A wait is woken by the subscription matcher when a message arrives on one of its streams: an event,
a person change, an internal event, a distinct_id repoint. A condition that compares against the
clock produces no message at all, so nothing wakes it. The periodic re-check is what has been
covering that gap, and it re-parks every wait in the fleet to do it.

Rejecting these at save time keeps that re-check a backstop rather than the mechanism: express the
same intent with a delay step and a condition that guards it, and the wait is woken by something
that actually happens.
"""

import re
from typing import Optional

from posthog.hogql import ast
from posthog.hogql.visitor import TraversingVisitor

# Zero-argument functions whose value advances with wall-clock time. A condition built on one of
# these changes truth value without anything happening, which is exactly what no stream can report.
CLOCK_FUNCTIONS = {"now", "today"}


class _ClockCallFinder(TraversingVisitor):
    def __init__(self) -> None:
        self.found: Optional[str] = None

    def visit_call(self, node: ast.Call) -> None:
        if self.found is None and node.name in CLOCK_FUNCTIONS:
            self.found = node.name
        super().visit_call(node)


def find_clock_function(condition_expr: ast.Expr) -> Optional[str]:
    """Name of the first clock function the expression calls, or None if it calls none."""
    finder = _ClockCallFinder()
    finder.visit(condition_expr)
    return finder.found


# The matcher keys its streams on person and distinct_id only, so a group change carries nothing it
# can look a parked job up by.
GROUP_FIELD_RE = re.compile(r"^group_\d+$")


class _GroupFieldFinder(TraversingVisitor):
    def __init__(self) -> None:
        self.found: Optional[str] = None
        # A lambda argument can shadow a group name, as in arrayExists(group_0 -> group_0 = 'pro', xs).
        self.bound: set[str] = set()

    def visit_lambda(self, node: ast.Lambda) -> None:
        added = {arg for arg in node.args if arg not in self.bound}
        self.bound |= added
        try:
            super().visit_lambda(node)
        finally:
            self.bound -= added

    def visit_field(self, node: ast.Field) -> None:
        if self.found is None and self._reads_group_property(node):
            self.found = str(node.chain[0])
        super().visit_field(node)

    def _reads_group_property(self, node: ast.Field) -> bool:
        # A group property is always read as group_<index>.properties.<key>, so a bare name is a
        # local or an unrelated identifier rather than the group the matcher cannot observe.
        if len(node.chain) < 2 or str(node.chain[1]) != "properties":
            return False
        root = str(node.chain[0])
        return root not in self.bound and bool(GROUP_FIELD_RE.match(root))


def find_group_field(condition_expr: ast.Expr) -> Optional[str]:
    """Name of the first group field the expression reads, or None if it reads none."""
    finder = _GroupFieldFinder()
    finder.visit(condition_expr)
    return finder.found

File: products/workflows/frontend/Workflows/hogflows/steps/waitClockConditions.ts
import { AnyPropertyFilter, PropertyFilterType } from '~/types'

/**
 * Detects a `wait_until_condition` condition that only the clock could ever satisfy.
 *
 * A wait is woken when a message arrives on one of its streams: an event, a person change, an
 * internal event, a distinct_id repoint. A condition that compares against the clock produces no
 * message at all, so no stream wakes it, and only the periodic re-check can advance it. The API
 * refuses it at save time. Finding it while the step is built is what turns that refusal into
 * guidance.
 *
 * Mirrors `find_clock_function` in
 * products/workflows/backend/services/wait_clock_conditions.py.
 */

// Zero-argument functions whose value advances with wall-clock time. A condition built on one of
// these changes truth value without anything happening, which is exactly what no stream can report.
const CLOCK_CALL = /(?:^|[^\w.])(now|today)\s*\(/

// Compiled from the author's filters rather than written by them, so comparing them would make an
// unchanged condition look edited. Mirrors `_DERIVED_FILTER_KEYS` in the API.
const DERIVED_FILTER_KEYS = ['bytecode', 'bytecode_error', 'source']

type ConditionFilters = {
    properties?: unknown[]
    events?: { properties?: unknown[] }[]
    actions?: { properties?: unknown[] }[]
} | null

type WaitCondition = { filters?: ConditionFilters } | null | undefined

/**
 * The expression with everything the parser does not read as code taken out: comments, and the text
 * of a plain string. An f-string keeps the expressions inside its braces, because the parser reads
 * those. Without this the search reports a call written in a comment or a literal, and misses one
 * written around a comment.
 */
function codeOnly(expression: string): string {
    let code = ''
    let index = 0

    while (index < expression.length) {
        const char = expression[index]

        if (char === '/' && expression[index + 1] === '*') {
            const end = expression.indexOf('*/', index + 2)
            index = end === -1 ? expression.length : end + 2
            code += ' '
        } else if (char === '-' && expression[index + 1] === '-') {
            const end = expression.indexOf('\n', index)
            index = end === -1 ? expression.length : end
            code += ' '
        } else if (char === "'" || char === '"' || char === '`') {
            const isTemplate = /(?:^|[^\w])[fF]$/.test(code)
            let body = ''
            index += 1
            while (index < expression.length && expression[index] !== char) {
                if (expression[index] === '\\') {
                    index += 1
                }
                body += expression[index] ?? ''
                index += 1
            }
            index += 1
            code += isTemplate ? ` ${templateExpressions(body)} ` : " '' "
        } else {
            code += char
            index += 1
        }
    }

    return code
}

/** The parts of an f-string body the parser evaluates, which is what sits between the braces. */
function templateExpressions(body: string): string {
    return (body.match(/\{[^{}]*\}/g) ?? []).join(' ')
}

/** Name of the first clock function the condition calls, or null if it calls none. */
export function findClockFunction(filters?: ConditionFilters): string | null {
    const propertyGroups = [
        filters?.properties,
        ...(filters?.events ?? []).map((entry) => entry?.properties),
        ...(filters?.actions ?? []).map((entry) => entry?.properties),
    ]

    for (const properties of propertyGroups) {
        for (const property of (properties ?? []) as AnyPropertyFilter[]) {
            // Only a HogQL expression carries a function call; every other filter type is a
            // key/operator/value the compiler renders without touching the clock.
            if (property?.type !== PropertyFilterType.HogQL || typeof property.key !== 'string') {
                continue
            }
            const match = CLOCK_CALL.exec(codeOnly(property.key))
            if (match) {
                return match[1]
            }
        }
    }

    return null
}

/** The parts of a wait condition a person wrote, with compiler output dropped. */
export function authoredCondition(condition: WaitCondition): unknown {
    const filters = condition?.filters
    if (!filters) {
        return condition
    }
    const authoredFilters = Object.fromEntries(
        Object.entries(filters).filter(([key]) => !DERIVED_FILTER_KEYS.includes(key))
    )
    return { ...condition, filters: authoredFilters }
}

export function clockConditionMessage(clockFunction: string): string {
    return (
        `This condition uses ${clockFunction}(), so it depends on the current time rather than on ` +
        'something happening, and it can only advance once the workflow checks it again. To wait for a point ' +
        'in time, use a delay step, then put this condition on the step after it.'
    )
}
